File: 1824-minimum-sideway-jumps/1824-minimum-sideway-jumps.cpp
class Solution {
public:
    int minSideJumps(vector<int>& obstacles) {
        map<pair<int,int>,int> dp;
        return solve(2, 0, obstacles, dp);
    }
    
    
    int solve(int pos, int i, vector<int>& obstacles, map<pair<int,int>,int> &dp){

               
        if(dp[{pos, i}])
            return dp[{pos, i}]; 
        int ans = INT_MAX;
        
        
        for(int k = i; k < obstacles.size() - 1; k++){
            if(pos == obstacles[k + 1]){
                for(int j = 1; j <= 3; ++j){
                    if(pos != j && j != obstacles[k])
                        ans = min(ans, 1 + solve(j, k, obstacles, dp));
                }
                dp[{pos, i}] = ans;
                return ans;
            }            
        }

            
        return 0;
    }
};

File: 1824-minimum-sideway-jumps/1824-minimum-sideway-jumps.java
class Solution {
    public int minSideJumps(int[] obstacles) {
        int []dp = new int[]{1, 0, 1};
        for(int obstacle : obstacles){
            if(obstacle > 0)
                dp[obstacle - 1] = 10000000;
            for(int i = 0; i < 3; ++i){
                if(i + 1 != obstacle)
                    dp[i]= Math.min(dp[i], Math.min(dp[(i + 1)%3], dp[(i + 2)%3]) + 1);
            }
        }
        return Math.min(dp[0], Math.min(dp[1], dp[2]));
    }
}